go/vt/vtgate/evalengine/arithmetic_expr.go

```diff
@@ -50,61 +50,60 @@ func (b *ArithmeticExpr) collation() collations.TypedCollation {
 	return collationNumeric
 }
 
-func (b *ArithmeticExpr) eval(env *ExpressionEnv) (EvalResult, error) {
-	lVal, err := b.Left.eval(env)
-	if err != nil {
-		return EvalResult{}, err
+func (b *ArithmeticExpr) eval(env *ExpressionEnv, out *EvalResult) {
+	var left, right EvalResult
+	left.init(env, b.Left)
+	right.init(env, b.Right)
+	if left.null() || right.null() {
+		out.setNull()
+		return
 	}
-	if lVal.null() {
-		return resultNull, nil
-	}
-	rVal, err := b.Right.eval(env)
-	if err != nil {
-		return EvalResult{}, err
-	}
-	if rVal.null() {
-		return resultNull, nil
-	}
-	if lVal.typ == querypb.Type_TUPLE || rVal.typ == querypb.Type_TUPLE {
+	if left.typeof() == querypb.Type_TUPLE || right.typeof() == querypb.Type_TUPLE {
 		panic("failed to typecheck tuples")
 	}
-	return b.Op.eval(lVal, rVal)
+	if err := b.Op.eval(&left, &right, out); err != nil {
+		throwEvalError(err)
+	}
 }
 
 // typeof implements the Expr interface
-func (b *ArithmeticExpr) typeof(env *ExpressionEnv) (querypb.Type, error) {
-	ltype, err := b.Left.typeof(env)
-	if err != nil {
-		return 0, err
-	}
-	rtype, err := b.Right.typeof(env)
-	if err != nil {
-		return 0, err
-	}
-	typ := mergeNumericalTypes(ltype, rtype)
-	return b.Op.typeof(typ), nil
+func (b *ArithmeticExpr) typeof(env *ExpressionEnv) querypb.Type {
+	/*
+		ltype, err := b.Left.typeof(env)
+		if err != nil {
+			return 0, err
+		}
+		rtype, err := b.Right.typeof(env)
+		if err != nil {
+			return 0, err
+		}
+		typ := mergeNumericalTypes(ltype, rtype)
+		return b.Op.typeof(typ), nil
+	*/
+	// TODO: make this less aggressive
```

Comment:
should this still be commented out?

Reply:
Yes, I've updated the comment to explain why this is still a TODO